clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -307,20 +331,33 @@ private MetadataCache handleMetadataResponse(MetadataResponse metadataResponse,
 
                     if (partitionMetadata.error().exception() instanceof InvalidMetadataException) {
                         log.debug("Requesting metadata update for partition {} due to error {}",
-                                new TopicPartition(metadata.topic(), partitionMetadata.partition()), partitionMetadata.error());
-                        requestUpdate();
+                                  new TopicPartition(metadata.topic(), partitionMetadata.partition()), partitionMetadata.error());
+                        success = false;
                     }
                 }
-            } else if (metadata.error().exception() instanceof InvalidMetadataException) {
-                log.debug("Requesting metadata update for topic {} due to error {}", metadata.topic(), metadata.error());
+            } else {
+                if (metadata.error().exception() instanceof InvalidMetadataException) {
+                    log.debug("Requesting metadata update for topic {} due to error {}", metadata.topic(), metadata.error());
+                    success = false;
+                }
+
+                if (metadata.error() == Errors.INVALID_TOPIC_EXCEPTION) {
+                    invalidTopics.add(metadata.topic());
+                } else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED) {
+                    unauthorizedTopics.add(metadata.topic());
```

Comment:
No, I believe because it's not considered a transient error.

Reply:
So you were right to notice an issue. While we weren't requesting immediate update before, the original behavior still would have included this topic in the next update, as it should, since ACLs can change. This is fixed.